libs/iplant-clojure-commons/src/clojure_commons/exception.clj

```diff
@@ -26,118 +22,189 @@
-                      exception
-                      (resp/bad-request (cheshire/encode exception)))))
-
-(defn bad-query-params-handler
-  [error error-type _]
-  (let [exception {:error_code ec/ERR_BAD_QUERY_PARAMETER
-                   :parameters (:parameters error-type)}]
-    (embed-error-info error
-                      exception
-                      (resp/bad-request (cheshire/encode exception)))))
+                     :reason (get-in hndlr-resp [:body :errors])}
+          response (-> hndlr-resp
+                       (assoc :body (cheshire/encode exception))
+                       (header/content-type "application/json; charset=utf-8"))]
+      (smw/log-response-with-exception error
+                                       request
+                                       response
+                                       exception)
+      response)))
+
+(defn- exception-with-reason
+  ([error-code reason]
+   (exception-with-reason error-code reason {}))
+  ([error-code reason m]
+   (assoc m
+     :error_code error-code
+     :reason     reason)))
+
+(defn- apply-headers
+  [response headers-map]
+  (reduce (fn [response [header-name header-value]]
+            (header/header response header-name header-value))
+          response headers-map))
+
+(defn default-error-handler
+  "Produces an error handler which generates exceptions with the
+  given error-code and HTTP response function."
+  [response-fn error-code & [header-map]]
+  (fn [error error-type request]
+    (let [reason    (:error error-type)
+          exception (exception-with-reason error-code reason (dissoc error-type :type :error))
+          response  (apply-headers (response-fn (cheshire/encode exception)) header-map)]
+      (smw/log-response-with-exception error request response exception)
+      response)))
+
+(defn temporary-redirect-handler
+  "Specialized exception handler for 302 responses. This could not be accomplished with the default
+   exception handler because best practices for 302 responses is an empty response body."
+  [error error-type request]
+  (let [reason (:error error-type)
+        exception (exception-with-reason ec/ERR_TEMPORARILY_MOVED reason (dissoc error-type :type :error))
+        location (:location error-type)
+        response (resp/found location)]
+    (smw/log-response-with-exception error request response exception)
+    response))
+
+(def authentication-not-found-handler
+  (default-error-handler resp/unauthorized ec/ERR_NOT_AUTHORIZED {"WWW-Authenticate" "Custom"}))
+
+(def bad-query-params-handler
+  (default-error-handler resp/bad-request ec/ERR_BAD_QUERY_PARAMETER))
+
+(def bad-request-field-handler
+  (default-error-handler resp/bad-request ec/ERR_BAD_OR_MISSING_FIELD))
+
+(def failed-dependency-handler
+  (default-error-handler resp/failed-dependency ec/ERR_MISSING_DEPENDENCY))
+
+(def forbidden-handler
+  (default-error-handler resp/forbidden ec/ERR_FORBIDDEN {"WWW-Authenticate" "Custom"}))
```

Comment:
Probably don't need the header since this is for "you're authenticated, you just aren't allowed to do that"